File: python/p0135.py
from typing import List

class Solution:
    def candy(self, ratings: List[int]) -> int:
        # If there's only one child, just give them 1 candy.
        if len(ratings) == 1:
            return 1
        
        # Initialize a list `candy` where each child gets at least one candy.
        # We start by giving 1 candy to every child.
        candy = [1] * len(ratings)
        
        # First pass: Traverse from left to right to ensure children with higher ratings than
        # the previous child get more candies than their left neighbor.
        for i in range(1, len(ratings)):
            if ratings[i] > ratings[i-1]:
                # If the current child's rating is higher than the previous child,
                # give the current child one more candy than the previous child.
                candy[i] = candy[i-1] + 1
        
        # Second pass: Traverse from right to left to ensure children with higher ratings than
        # the next child get more candies than their right neighbor.
        for i in range(len(ratings) - 2, -1, -1):
            if ratings[i] > ratings[i+1]:
                # If the current child's rating is higher than the next child,
                # make sure they have more candies than the next child.
                # Only update if the current child doesn't already have enough candies.
                if candy[i] <= candy[i+1]:
                    candy[i] = candy[i+1] + 1
        
        # The total number of candies needed is the sum of the `candy` list.
        return sum(candy)

File: javascript/p0135.js
/**
 * @param {number[]} ratings
 * @return {number}
 */
var candy = function (ratings) {
  // If there's only one child, return 1 candy
  if (ratings.length === 1) {
    return 1;
  }

  // Initialize an array `candies` where each child starts with 1 candy
  let candies = new Array(ratings.length).fill(1);

  // First pass: Left to right
  // If the current child has a higher rating than the previous child,
  // give them more candies than the previous child
  for (let i = 1; i < ratings.length; i++) {
    if (ratings[i] > ratings[i - 1]) {
      candies[i] = candies[i - 1] + 1;
    }
  }

  // Second pass: Right to left
  // If the current child has a higher rating than the next child,
  // make sure they have more candies than the next child
  for (let i = ratings.length - 2; i >= 0; i--) {
    if (ratings[i] > ratings[i + 1]) {
      candies[i] = Math.max(candies[i], candies[i + 1] + 1);
    }
  }

  // The total number of candies needed is the sum of the `candies` array
  return candies.reduce((sum, candy) => sum + candy, 0);
};
